### DISCUSSION TITLE
Blazor with CancellationToken support

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#96] [General]
# Blazor with CancellationToken support

What happens when a user navigates away and still has a pending request to our server? Or what happens if we have a request which might take forever and wastes our resources?
Let's tackle these problems with a CancellationToken.

https://steven-giesel.com/blogPost/69051bce-81b6-410c-9b9e-c2a0f1c88371

**Comment by kjkrum** [comment]
`NavigationManager` seems to live in an enclosing scope. If you add an event handler to `NavigationManager.LocationChanged` in `OnInitializedAsync`, then every time you navigate to this component, a new handler will be added. The handler may leak other members of your component, like your CTS.

I discovered this by setting a breakpoint in my handler. Each time I visit the component, the breakpoint is hit as many times as I've visited the component.

**Reply by linkdotnet** [reply]
Good point - on dispose one should detach from the event anyway otherwise the component can’t be garbage collected.
That also should fix the behaviour you are seeing. Bonus points for disposing the CTS

**Comment by valuator18** [comment]
It's probably a better pattern to have your component implement `IDisposable` and cancel the token in the `Dispose()` method. Components will get disposed automatically when the user navigates, negating the need for a `NavigationManager` or event handlers.

**Reply by linkdotnet** [reply]
That is another (probably, as you said) better option. 

I feel like I had a reason to use the `NavigationManager` instead of `IDisposable`, but I also can't remember anymore :D
